monai/apps/auto3dseg/hpo_gen.py

```diff
@@ -0,0 +1,263 @@
+
+        Args:
+            output_folder
+        """
+        task_id = self.get_task_id()
+        task_prefix = os.path.basename(self.algo.get_output_path())
+        write_path = os.path.join(output_folder, task_prefix + task_id)
+        self.obj_filename = os.path.join(write_path, "algo_object.pkl")
+
+        if isinstance(self.algo, BundleAlgo):
+            self.algo.export_to_disk(output_folder, task_prefix + task_id)
+            algo_export_to_pickle(self.algo, self.obj_filename, self.algo_templates_dir)
+        else:
+
+            ConfigParser.export_config_file(self.params, write_path)
+            logger.info(write_path)
+
+    def set_score(self, acc):
+        """
+        Report the acc to nni server
+        """
+        if has_nni:
+            nni.report_final_result(acc)
+        return
+
+    def run_algo(self, obj_filename: str, output_folder: str = ".", algo_templates_dir=None) -> None:  # type: ignore
+        """
+        The python interface for NNI to run
+
+        Args:
+            obj_filename: the pickle-exported Algo object.
+            output_folder: the root path of the algorithms templates.
+            algo_templates_dir: the algorithm_template. It must contain algo.py in the follow path:
+                ``{algorithm_templates_dir}/{network}/scripts/algo.py``
+        """
+        if not os.path.isfile(obj_filename):
+            raise ValueError(f"{obj_filename} is not found")
+
+        self.algo, self.algo_templates_dir = algo_import_from_pickle(obj_filename, algo_templates_dir)
+
+        # step1 sample hyperparams
+        params = self.get_hyperparameters()
+        # step 2 update model
+        self.update_params(params)
+        # step 3 generate the folder to save checkpoints and train
+        self.generate(output_folder)
+        self.algo.train(self.params)
+        # step 4 report validation acc to controller
+        acc = self.algo.get_score()
+        self.set_score(acc)
+
+    def get_history(self, output_folder: str = ".", algo_templates_dir=None) -> List:  # type: ignore
+        """
+        Get the history of the bundleAlgo object with their names/identifiers
+
+        Args:
+            output_folder: the root path of the algorithms templates.
+            algo_templates_dir: the algorithm_template. It must contain algo.py in the follow path:
+                ``{algorithm_templates_dir}/{network}/scripts/algo.py``
+        """
+
+        history = []
+
+        for task_name in os.listdir(output_folder):
+            write_path = os.path.join(output_folder, task_name)
+
+            if not os.path.isdir(write_path):
+                continue
+
+            obj_filename = os.path.join(write_path, "algo_object.pkl")
+            if not os.path.isfile(obj_filename):  # saved mode pkl
+                continue
+
+            algo, _ = algo_import_from_pickle(obj_filename, algo_templates_dir)
+            history.append({task_name: algo})
```

Comment:
I think we should `append` in the `generate` progress instead of the `get_history`, refer to:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/apps/auto3dseg/bundle_gen.py#L459

Thanks.

Reply:
The `generate` function is part of run_algo and will be completely maintained by `nni` in the runtime. We will get nothing from nni except files and folders. That's why I have to check the file system and get the history here.